core/thread/inc/TPool.h

```diff
@@ -0,0 +1,204 @@
+#ifndef TPOOL_H_
+#define TPOOL_H_
+
+#include "TCollection.h"
+#include "TObjArray.h"
+#include "ROOT/TSeq.h"
+
+template<class subc>
+class TPool {
+public:
+   explicit TPool(){};
+   explicit TPool(size_t nThreads){};
+
+   // // Map
+   // //these late return types allow for a compile-time check of compatibility between function signatures and args,
+   // //and a compile-time check that the argument list implements a front() method (all STL sequence containers have it)
+   template<class F> auto Map(F func, unsigned nTimes) -> std::vector<decltype(func())>;
+   template<class F, class T> auto Map(F func, T &args) -> std::vector < decltype(++(args.begin()), args.end(), func(args.front())) >;
```

Comment:
Couldn't find a way to avoid the decltype(++(args.begin()), args.end(), func(args.front()) if we want to check if the object is iterable.

Tried the [enable_if hack](https://github.com/root-mirror/root/pull/148/files/fcb4aadd6967e1d7ea831aa4159d36a485722a85#r58660007) here, but accessing T::begin() and T::end() requires object creation.